src/librustc_mir/interpret/operator.rs

```diff
@@ -371,29 +374,45 @@ impl<'mir, 'tcx, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
                     Not => !val,
                     _ => bug!("Invalid bool op {:?}", un_op),
                 };
-                Ok(ImmTy::from_scalar(Scalar::from_bool(res), self.layout_of(self.tcx.types.bool)?))
+                Ok((Scalar::from_bool(res), false, self.tcx.types.bool))
             }
             ty::Float(fty) => {
                 let res = match (un_op, fty) {
                     (Neg, FloatTy::F32) => Scalar::from_f32(-val.to_f32()?),
                     (Neg, FloatTy::F64) => Scalar::from_f64(-val.to_f64()?),
                     _ => bug!("Invalid float op {:?}", un_op),
                 };
-                Ok(ImmTy::from_scalar(res, layout))
+                Ok((res, false, layout.ty))
             }
             _ => {
                 assert!(layout.ty.is_integral());
                 let val = self.force_bits(val, layout.size)?;
-                let res = match un_op {
-                    Not => !val,
+                let (res, overflow) = match un_op {
+                    Not => (self.truncate(!val, layout), false), // bitwise negation, then truncate
                     Neg => {
+                        // arithmetic negation
                         assert!(layout.abi.is_signed());
-                        (-(val as i128)) as u128
```

Comment:
This, too, was wrong, I think. `val` can't just be cast to `i128`, it needs to be sign-extended.

But it [looks like](https://play.rust-lang.org/?version=nightly&mode=release&edition=2018&gist=abf2a0b0da922df799b45bb0d5d96a36) the computed result is correct nevertheless.

Fun fact: looks like the MIR of a `const` body contains overflow checks even when building in release mode! Is that deliberate? That makes testing these things even harder than it already is anyway...

Reply:
Also, I wasn't sure of the test coverage of all of these arithmetic operations here; I didn't want to spend time adding more tests if there already are some. Where would I find those?

Reply:
I see bunch of tests in `src/tests/ui/consts/const-eval/`, specifically the `const-eval-overflow-*` tests. There's also some incidental coverage in `src/tests/mir-opt/const-prop`.

Reply:
I couldn't find anything that checked the overflow behavior of unary/binary operators (as opposed to the `overflowing_`/`wrapping_` methods). So I added a test for that.